kedro/io/memory_data_set.py

```diff
@@ -121,3 +121,18 @@ def _exists(self) -> bool:
         if self._data is None:
             return False
         return True
+
+    def set_max_loads(self, max_loads):
+        """Set how many times this dataset can be loaded before its data being cleared to release
```

Comment:
Would like to avoid the counting if possible.

Could we perhaps create an interface that allows each data set to be used as a python context, e.g.,

```python
with memory_ds.use(): # something like this
    memory_ds.load()
    # process, load some more...
```

Then the disposal would happen automatically, we could reconfigure the pipeline to use these contexts when "accessing" data sources.

**Key benefit**: User doesn't have to do annoying counting to figure out how many times a data set will be called in the pipeline.

Reply:
The interface is a bit clunky but the target `users` here are the runners and for them this interface is the simplest to use. Integrating context manager calls into the runners, particularly releasing the contexts mid way through the run, would be quite complicated.